doc/chapters/api/index.rst

```diff
@@ -3,131 +3,230 @@
+All methods in this API can and will return D-Bus errors on failure. All calls that don't return a
+D-Bus error have worked successfully. Any call that returns a D-Bus error has failed. All calls that
+change a state and raise a D-Bus error shall have their side effects documented in this chapter.
+
+For each method, we describe the various error types and their potential sources. Some of these
+errors have a note saying that they put the container in an `invalid state`. An invalid state means
+that no other operations can be reliably done, and SoftwareContainer will disallow any operations on
+a container that is in an invalid state, except for destroying them.
 
 Methods
 -------
 
 BindMount
 ---------
-Binds a path on the host to the container. Any missing parent directories in
-the path will be created.
-
-:Parameters:
-        :containerID: ``int32`` The ID obtained by CreateContainer method.
-        :pathInHost: ``string`` absolute path in the host.
-        :pathInContainer: ``string`` the absolute path to mount to in the container.
-        :readOnly: ``bool`` indicates whether the directory is read-only or not.
-
-|
+Binds a path on the host to the container. Any missing parent directories in the path will be
+created.
+
+Parameters
+##########
+* containerID: ``int32`` The ID obtained by CreateContainer method.
+* pathInHost: ``string`` absolute path in the host.
+* pathInContainer: ``string`` the absolute path to mount to in the container.
+* readOnly: ``bool`` indicates whether the directory is read-only or not.
+
+Error sources
+#############
+* Invalid ID: No matching container exists.
+* Bad state: The container is not started or ready
+* Bad path
+    * The host path does not exist
+    * The destination path is not absolute
+    * The destination path is already mounted to
+* Filesystem issues
+    * Could not create parent directories
+    * Could not create the target directory (if mounting a directory)
+    * Could not touch the target file (if mounting a file)
+* Mount issues
+    * Mount failed (:linuxman:`mount(2)` failed)
+    * Couldn't move the mount
+    * Failed to re-mount read-only
+
+Currently, failing mounts are not rolled back.
 
 Create
 ------
 Creates a container with given configuration.
 
-:Parameters:
-        :config: ``string`` config file in json format.
+Parameters
+##########
+* config: ``string`` config file in json format.
 
-                Example config JSON::
+Example config JSON::
 
-                [{"enableWriteBuffer": true}]
+[{"enableWriteBuffer": true}]
 
-|
+Return Values
+#############
+* containerID: ``int32`` ID of created SoftwareContainer.
 
-:Return Values:
-        :containerID: ``int32`` ID of created SoftwareContainer.
+Error sources
+#############
+* Network setup issues
+    * Bridge device not available and unable to set it up
+    * Given netmask can not be used to assign IP address to container
+    * See :ref:`Network setup<network-setup>`
```

Comment:
I don't think it would hurt to add this type of reference to setup instructions in more sections.